### PR TITLE
Guzzle Authorization Middleware

### PR DESCRIPTION
**Pull request by frankverhoeven**

### CODE REVIEW

File: src/Token.php [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+     */
+    public function getExpiresAt()
+    {
+        return $this->expiresAt;
+    }
+
+    public function isExpired(): bool
+    {
+        if (null === $this->expiresAt) {
+            return false;
```
Comment (Feijs) [inline src/Token.php]: Should it be possible to have an infinitely valid token?

File: src/TokenManager/CachedToken.php [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+    public function __construct(CacheItemPoolInterface $cachePool, TokenManagerInterface $innerTokenManager)
+    {
+        $this->cachePool = $cachePool;
+        $this->innerTokenManager = $innerTokenManager;
+    }
+
+    public function getToken(): Token
+    {
+        $item = $this->cachePool->getItem(self::class);
```
Comment (Feijs) [inline src/TokenManager/CachedToken.php]: This will store a single token based on this class name? How will that work for tokens for different tenant ids?

**Review by Feijs — COMMENTED** [review]
Since this is a public repo, I think a license would be advisable. See https://github.com/MyOnlineStore/skeleton for that and some more useful files to include

File: src/TokenManager/CachedToken.php (hunk 2 above)
Comment (frankverhoeven) [reply src/TokenManager/CachedToken.php]: The `cachePool` instance provided will have to scope for user/store ID. Since that info is not available here, there is no way to add it right here.

Comment (Feijs) [reply src/TokenManager/CachedToken.php]: Ah right :+1:

File: src/Token.php (hunk 1 above)
Comment (frankverhoeven) [reply src/Token.php]: Depends on the kind of token. I chose this solution to keep things as generic as possible. I could agree on enforcing an expiry time as well though.

File: src/TokenManager/CachedToken.php (hunk 2 above)
Comment (frankverhoeven) [reply src/TokenManager/CachedToken.php]: Another possible solution would be to add an interface that provides the ID, which has to be implemented by the user. Like `UriProviderInterface`.

Actually, I could probably just use the URI provided by that implementation as well.

File: src/Token.php (hunk 1 above)
Comment (Feijs) [reply src/Token.php]: > I chose this solution to keep things as generic as possible

Perhaps you could introduce a (constructor) parameter on whether to allow tokens without expiry?

File: src/TokenManager/CachedToken.php (hunk 2 above)
Comment (Feijs) [reply src/TokenManager/CachedToken.php]: > Actually, I could probably just use the URI provided by that implementation as well.

Indeed, but using the tenant specific cache would be OK as well

Comment (frankverhoeven) [reply src/TokenManager/CachedToken.php]: > Indeed, but using the tenant specific cache would be OK as well

Well, an implementation for `UriProviderInterface` is required for the JWT manager, so it will be available anyway. By using it here to scope the cached token, we can guarantee it is saved correctly without depending on the `cachePool` implementation. I actually like that idea better than being dependant on an external implementation.

TL;DR, I will add `UriProviderInterface` as a dependency and use the uri to scope the cached token.

File: src/Token.php (hunk 1 above)
Comment (frankverhoeven) [reply src/Token.php]: What would be the added benefit of an additional parameter, instead of just using the nullable expiresAt field?

Comment (Feijs) [reply src/Token.php]: That you keep the "generic" solution: that the package supports having infinitely valid tokens, yet also offers the possibility of blocking them.

What I meant was something like:
```
/**
 * @var bool
 */
private $allowNonExpiringTokens;

//....
if (null === $this->expiresAt) {
      return !$this->allowNonExpiringTokens;
}
```

**Review by Feijs — APPROVED** [review]
